### PR TITLE
Aws/snsverify

### PR DESCRIPTION
**Pull request by jonathangold**
Here are the methods to verify sns message signatures.

### CODE REVIEW

File: resources/aws_ec2.go [hunk 1]
```diff
@@ -971,25 +1009,28 @@ func (obj *AwsEc2Res) snsListener(listenAddr string) (net.Listener, error) {
...
-		}:
-		case <-obj.closeChan:
-		}
+		log.Printf("%s: error decoding post: %s", obj, err)
 		return
 	}
+	// Verify the x509 signature. If there is an error verifying the
+	// signature, we print the error, ignore the event and return.
+	if err := obj.snsVerifySignature(post); err != nil {
+		http.Error(w, "Bad request", http.StatusBadRequest)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: put log message first please

File: resources/aws_ec2.go [hunk 2]
```diff
@@ -1020,7 +1061,7 @@ func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
 		if err != nil {
 			select {
 			case obj.awsChan <- &chanStruct{
-				err: errwrap.Wrapf(err, "error confirming subscription"),
+				err: errwrap.Wrapf(err, "errpr processing event"),
```
Comment (purpleidea) [inline resources/aws_ec2.go]: typo

File: resources/aws_ec2.go [hunk 3]
```diff
@@ -69,6 +75,8 @@ const (
 	SnsPolicyService = "events.amazonaws.com"
 	// SnsPolicyAction is the specific permission we are granting in the policy.
 	SnsPolicyAction = "SNS:Publish"
+	// SnsCertURLRegex is used to make sure we only download certificates from amazon.
```
Comment (purpleidea) [inline resources/aws_ec2.go]: add a description describing in human the pattern we're looking for please.

File: resources/aws_ec2.go [hunk 4]
```diff
@@ -971,25 +1009,28 @@ func (obj *AwsEc2Res) snsListener(listenAddr string) (net.Listener, error) {
...
 		return
 	}
-	// decode json
+	// Decode the post. If we are unable to decode the data, we print the
+	// error, ignore the post and return.
 	decoder := json.NewDecoder(req.Body)
 	var post postData
 	if err := decoder.Decode(&post); err != nil {
 		http.Error(w, "Bad request", http.StatusBadRequest)
-		select {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: let's keep this in but wrap it in a condition some sort of something ErrorXXXStrict

File: resources/aws_ec2.go [hunk 5]
```diff
@@ -85,6 +93,29 @@ const (
 	waitTimeout = 400
 )
 
+// snsCanonicalTemplate is a template that formats sns messages according to
+// amazon's canonical message format needed to check message signatures.
+// We use printf for the URL so we don't escape special characters.
+const snsCanonicalTemplate = `Message
```
Comment (purpleidea) [inline resources/aws_ec2.go]: sadly I think this is a good use case for:

`var str string
str += fmt.Sprintf(XXX)
if foo {
   str += fmt.Sprintf(XXX)

}
```

Also please add a link to the docs about this so people don't think we're insane. (amazon is insane)

**Comment by jonathangold** [comment]
@purpleidea See comment in IRC re `fmt.Sprintf` vs `buffer.Write` before you merge.

File: resources/aws_ec2.go [hunk 6]
```diff
@@ -1038,6 +1075,91 @@ func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
...
+	}
+	return parsedCert, nil
+}
+
+// snsCanonicalFormat formats post messages as required for signature
+// verification. For more information about this requirement see:
+// http://docs.aws.amazon.com/sns/latest/dg/SendMessageToHttp.verify.signature.html
+func (obj *AwsEc2Res) snsCanonicalFormat(post postData) []byte {
+	var str string
+	str += fmt.Sprintf("Message\n")
```
Comment (purpleidea) [inline resources/aws_ec2.go]: A lot of these can be simplified, eg: 

` 	str += "Message\n"`

or 

`str += post.MessageID + "\n"` (if it's already a string)

or 

`str += string(post.MessageID) + "\n"` (if it isn't)

File: resources/aws_ec2.go [hunk 7]
```diff
@@ -1038,6 +1075,91 @@ func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
...
+	str += fmt.Sprintf("%s\n", post.Timestamp)
+	if post.Token != "" {
+		str += fmt.Sprintf("Token\n")
+		str += fmt.Sprintf("%s\n", post.Token)
+	}
+	str += fmt.Sprintf("TopicArn\n")
+	str += fmt.Sprintf("%s\n", post.TopicArn)
+	str += fmt.Sprintf("Type\n")
+	str += fmt.Sprintf("%s\n", post.Type)
+	return []byte(str)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Assuming you want the function signature to stay the same, I think this is fine as is, modulo the above mentioned changes.

File: resources/aws_ec2.go [hunk 8]
```diff
@@ -69,6 +73,10 @@ const (
 	SnsPolicyService = "events.amazonaws.com"
 	// SnsPolicyAction is the specific permission we are granting in the policy.
 	SnsPolicyAction = "SNS:Publish"
+	// SnsCertURLRegex is used to make sure we only download certificates
+	// from amazon. This regex will match "https://sns.***.amazonaws.com/"
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Either comment or regex is wrong since there is a dash (`-`) you didn't talk about.

File: resources/aws_ec2.go [hunk 9]
```diff
@@ -133,6 +141,12 @@ type AwsEc2Res struct {
 
 	WatchEndpoint   string `yaml:"watchendpoint"`   // the public url of the sns endpoint, eg: http://server:12345/
 	WatchListenAddr string `yaml:"watchlistenaddr"` // the local address or port that the sns listens on, eg: 10.0.0.0:23456 or 23456
+	// ErrorJSONStrict is a bool that controls weather json.Decode(post)
+	// errors are sent to the engine or not. If it is set to true and an
+	// error is produced, Watch() will return the error and the engine will
+	// shut down.
+	ErrorJSONStrict bool `yaml:"errorjsonstrict"`
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Let's call this something without the word JSON. (The implementation should be allowed to change without the interface needing to.) How about ErrorOnInvalidMessages or similar (replace Messages with whatever they're called)

File: resources/aws_ec2.go [hunk 10]
```diff
@@ -971,25 +995,38 @@ func (obj *AwsEc2Res) snsListener(listenAddr string) (net.Listener, error) {
...
 		return
 	}
-	// decode json
+	// Decode the post. If an error is produced we either ignore the post,
+	// or if ErrorJSONStrict is true, send the error through awsChan so
+	// Watch() can return the error and the engine can shut down.
 	decoder := json.NewDecoder(req.Body)
 	var post postData
 	if err := decoder.Decode(&post); err != nil {
-		http.Error(w, "Bad request", http.StatusBadRequest)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Why did you move this?
My comment about having log earlier was about log vs. http -- not "internal action" vs. "http". I think the http should happen first, and then we shutdown. Yes you can put the log message before that.

File: resources/aws_ec2.go [hunk 11]
```diff
@@ -971,25 +995,38 @@ func (obj *AwsEc2Res) snsListener(listenAddr string) (net.Listener, error) {
...
+		if obj.ErrorJSONStrict {
+			select {
+			case obj.awsChan <- &chanStruct{
+				err: errwrap.Wrapf(err, "error decoding incoming POST, check struct formatting"),
+			}:
+			case <-obj.closeChan:
+			}
+			return
 		}
+		log.Printf("%s: error decoding post: %s", obj, err)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Don't these two need to be moved up to the top? If ErrorJSONStrict we don't get the log message!

File: resources/aws_ec2.go (hunk 10 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: I figured we wanted one or the other, but that makes sense. Changed.

File: resources/aws_ec2.go (hunk 11 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: I figured we wanted one or the other, but this makes sense. Changed.

**Comment by jonathangold** [comment]
@purpleidea Changes pushed.

File: resources/aws_ec2.go [hunk 12]
```diff
@@ -133,6 +142,12 @@ type AwsEc2Res struct {
 
 	WatchEndpoint   string `yaml:"watchendpoint"`   // the public url of the sns endpoint, eg: http://server:12345/
 	WatchListenAddr string `yaml:"watchlistenaddr"` // the local address or port that the sns listens on, eg: 10.0.0.0:23456 or 23456
+	// ErrorOnMalformedPost controls whether or not malformed HTTP post
+	// requests, that cause JSON decoder errors, will also make the engine
+	// shut down. If ErrorOnBadRequest set to true and an error occurs,
```
Comment (purpleidea) [inline resources/aws_ec2.go]: ErrorOnBadRequest ??? What's that?

File: resources/aws_ec2.go [hunk 13]
```diff
@@ -1038,6 +1075,91 @@ func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
...
+	}
+	// convert the message to canonical form
+	message := obj.snsCanonicalFormat(post)
+	// decode the message signature from base64
+	signature, err := base64.StdEncoding.DecodeString(post.Signature)
+	if err != nil {
+		return errwrap.Wrapf(err, "error decoding string")
+	}
+	// check the signature against the message
+	err = cert.CheckSignature(x509.SHA1WithRSA, message, signature)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: err := and group with the if please.

alternatively we could

return errwrap.Wrapf(cert.CheckSignature(x509.SHA1WithRSA, message, signature),  "error checking signature")

but that's too long and messy. Better to do the first fix and keep it obvious for crypto.

File: resources/aws_ec2.go [hunk 14]
```diff
@@ -1038,6 +1075,91 @@ func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
...
+	if err != nil {
+		return nil, errwrap.Wrapf(err, "http get error")
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errwrap.Wrapf(err, "error reading post body")
+	}
+	// decode and the certificate and discard the rest
+	decodedCert, _ := pem.Decode(body)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: what's the second arg? an error?

Comment (jonathangold) [reply resources/aws_ec2.go]: It's any junk data that may exist in the response after the pem certificate. I'll make that more clear in the comment.

File: resources/aws_ec2.go (hunk 12 above)
Comment (jonathangold) [reply resources/aws_ec2.go]: whoops

**Comment by jonathangold** [comment]
@purpleidea done.
